Support for setting user timezone and display local time in popover

hackerkid opened the pull request:
- [X] Add option for setting timzone in user settings.
- [X] Show local time of user in user_popover.
- [x] Show the time in the correct format 12 / 24 hour as per user settings
- [x] Set user timezone automatically during signup

Fixes https://github.com/zulip/zulip/issues/1506

**UI for setting the Time Zone**
![screenshot from 2017-04-03 03 06 16](https://cloud.githubusercontent.com/assets/7190633/24591239/b692b568-181a-11e7-8d1f-0b45fba1c4c2.png)

**Local time of user being shown in the popover**
![screenshot from 2017-04-03 00 41 36](https://cloud.githubusercontent.com/assets/7190633/24590180/59ce2bfa-1806-11e7-9a4b-133703388f96.png)

showell commented on static/js/people.js:
```diff
@@ -178,6 +178,13 @@ exports.get_full_name = function (user_id) {
     return people_by_user_id_dict.get(user_id).full_name;
 };
 
+exports.get_user_time_by_email = function (email) {
```
Can we not make this lookup work by userid?

hackerkid replied:
Fixed :)

showell commented:
I'm slightly against adding an `if_equal` helper to our templates.  We generally try to keep our templates kind of dumb, and if we need to generate true/false values, we can just make them be part of the context.  This is generally easier to write tests for.  I'm curious to hear what @timabbott says, though, as in this particular case, `if_equal` is pretty innocuous.

timabbott commented on zproject/settings.py:
```diff
@@ -809,6 +809,8 @@ def get_secret(key):
             'third/jquery-perfect-scrollbar/js/perfect-scrollbar.js',
             'third/lazyload/lazyload.js',
+            'node_modules/moment/moment.js',
+            'third/moment-timezone/moment-timezone-with-data.min.js',
```
Does `third/moment-timezone/moment-timezone-with-data.min.js` not exist in NPM?  That'd be the preferred way to do it.  If that isn't possible, let's check in the non-minified version of this library -- our tools will minify it, and importantly the Git history will be much more useful than if we check in a minified third-party thing.

timabbott replied:
If we are checking it in, we should update docs/THIRDPARTY.md

hackerkid replied:
Changed to NPM.

timabbott commented on static/js/server_events.js:
```diff
@@ -293,6 +293,9 @@ function dispatch_normal_event(event) {
             page_params.default_language = event.setting;
         }
+        if (event.setting_name === 'timezone') {
+            page_params.timezone = event.setting;
```
This commit's message mentions a "timzone", which sounds like a pretty awesome, but probably should be changed to a "timezone" :)

hackerkid replied:
Changed the commit message. But the branch name is still timzone :p

timabbott commented on static/js/people.js:
```diff
@@ -178,6 +178,16 @@ exports.get_full_name = function (user_id) {
+exports.get_user_time = function (user_id) {
+    var user_timezone = people.get_person_from_user_id(user_id).timezone;
+    if (user_timezone) {
+        if(page_params.twenty_four_hour_time) {
```
Add space after the `if`.

timabbott commented on static/third/moment-timezone/moment-timezone-with-data.min.js:
```diff
@@ -0,0 +1,7 @@
+//! Copyright (c) JS Foundation and other contributors
+//! license : MIT
+//! github.com/moment/moment-timezone
+!function(a,b){"use strict";"function"==typeof define&&define.amd?define(["moment"],b):"object"==typeof module&&module.exports?module.exports=b(require("moment")):b(a.moment)}(this,function(a){"use strict";function b(a){return a>96?a-87:a>64?a-29:a-48}function c(a){var c,d=0,e=a.split("."),f=e[0],g=e[1]||"",h=1,i=0,j=1;for(45===a.charCodeAt(0)&&(d=1,j=-1),d;d<f.length;d++)c=b(f.charCodeAt(d)),i=6 ...
```
You can remove this from the commit adding `moment` now, right?

timabbott replied:
We don't need to check this in, right?

hackerkid replied:
Fixed :)

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -966,7 +966,7 @@ def test_change_timezone(self):
-            events = self.do_test(lambda: do_change_timezone(self.user_profile, setting_value))
+            events = self.do_test(lambda: do_change_timezone(self.user_profile, setting_value), num_events=2)
             error = schema_checker('events[0]', events[0])
             self.assert_on_error(error)
```
Add a schema checker check for the new event?

hackerkid replied:
Added the schema checker.

hackerkid commented:
@timabbott All tests are passing now.

timabbott commented:
@hackerkid nice!  I merged all of the commits except that the one that adds the setting UI.  For that one, there are 2 issues:
(1) Merge conflicts with our having rearranged a lot of the settings JS code into separate files.
(2) The CSS for the setting display could use some work; I think maybe we want a structure where the label sits above the dropdown menu?  The alignment certainly feels off:

![image](https://cloud.githubusercontent.com/assets/2746074/24774984/f0cbaa34-1acf-11e7-8201-540084e97b37.png)

Can you fix those up?  Thanks @hackerkid! :)

hackerkid commented:
@timabbott Changed the dropdown to bottom of label. I also removed the addition of `if_equal` helper to handlebar.

![screenshot from 2017-04-08 21 30 52](https://cloud.githubusercontent.com/assets/7190633/24830388/b7324b0a-1ca2-11e7-84f4-f13c36cb357d.png)

hackerkid commented:
@timabbott I also added a commit to set user timezone automatically during signup. For that I moved moment and moment-timezone from app to common in JS_SPECS.

hackerkid commented:
@timabbott Please review when you are free (In case you missed this PR).

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2231,6 +2233,16 @@ def do_set_user_display_setting(user_profile, setting_name, setting_value):
+               active_user_ids(user_profile.realm))
+    if user_profile.is_bot:
+        send_event(dict(type='realm_bot', op='update', bot=payload),
+                   bot_owner_userids(user_profile))
```
We don't need an event at all for bots, and we only need the `realm_user` event if it's a timezone change, not some other display setting change.

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -975,6 +975,37 @@ def test_user_default_language(self):
+        # type: () -> None
+        """
+        Check if the default language of new user is the default language
+        of the realm.
```
This description is copied from another test.

timabbott commented:
@hackerkid I apparently failed to merge the things I'd thought I had.  OK, so a few issues with this version:
* Bots don't need timezones.
* A few other things I commented on above.

I fixed some of those issues and merged all but the last commit (for real this time!).  For finishing that, you should fix:
* the new signup test (might actually be better to just add a few lines to an existing test about timezone rather than adding a totally new one)
* The `zerver/views/registration.py` code is not quite optimal -- ideally, it should set the timezone via `do_create_user`, rather than `do_set_user_display_setting`, since then we avoid an unnecessary timezone-change event on every new user signup.  Though we'll still need to use the `do_set_user_display_setting` approach after `do_change_full_name` in the `do_activate_user` code path.

Thanks @hackerkid!

hackerkid commented:
@timabbott Ready for another review.

timabbott commented:
I fixed one indentation issue, and merged this.  Thanks @hackerkid!
